rabbitmq/rabbitClient.go

```diff
@@ -12,71 +12,95 @@ const (
 )
 
 type rabbitMqClient struct {
-	url string
+	url    string
+	pubMut sync.Mutex
 
 	conn *amqp.Connection
 	ch   *amqp.Channel
 
 	connErrCh chan *amqp.Error
 	chanErr   chan *amqp.Error
+	arkCh     chan uint64
+	narkCh    chan uint64
 }
 
 // NewRabbitMQClient creates a new rabbitMQ client instance
-func NewRabbitMQClient(ctx context.Context, url string) (*rabbitMqClient, error) {
+func NewRabbitMQClient(url string) (*rabbitMqClient, error) {
 	rc := &rabbitMqClient{
-		url: url,
+		url:    url,
+		pubMut: sync.Mutex{},
 	}
 
 	err := rc.connect()
 	if err != nil {
 		return nil, err
 	}
 
-	go rc.connListener(ctx)
-
 	return rc, nil
 }
 
 // Publish will publich an item on the rabbitMq channel
 func (rc *rabbitMqClient) Publish(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
-	return rc.ch.Publish(
-		exchange,
-		key,
-		mandatory,
-		immediate,
-		msg,
-	)
-}
-
-// dial will return a rabbitMq connection
-func (rc *rabbitMqClient) dial(url string) (*amqp.Connection, error) {
-	return amqp.Dial(url)
-}
-
-func (rc *rabbitMqClient) connListener(ctx context.Context) {
+	rc.pubMut.Lock()
+	defer rc.pubMut.Unlock()
+
+	var err error
+
+	// In order to avoid losing any event, check rabbitmq ack event for the
+	// published message. If not-acknowledged, check if there is a connection or
+	// channel issue, and after that is solved try again.  This was done to
+	// make sure no event is lost, for example, if rabbitmq connection is not
+	// closing gracefully (port disabled from firewall), it may happen that the
+	// main loop will not catch the conn err event, and it will still try to
+	// publish the message.
 	for {
+		err = rc.ch.Publish(
+			exchange,
+			key,
+			mandatory,
+			immediate,
+			msg,
+		)
+
 		select {
+		case <-rc.arkCh:
+			log.Debug("Publish: published message ack")
```

Comment:
can we do a return here instead of multiple continues on the other cases and a break?